istft: Use unfold_backward instead of col2im

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88061
* __->__ #88060

`unfold_backward` implements the same operation as `col2im` but without support
for 2d kernels or dilation. However, `istft` doesn't use any of those features
and `unfold_backward` actually has a faster `TensorIterator` based
implementation so we should use it here instead.

In the example from #87353 I see a 2x speedup on both CPU and CUDA.

On a wider variety of sizes and inputs I still see speedups across the board, especially 
on CPU since `col2im` isn't parallelized but `unfold_backward` is:

| device | shape           | hop_length | Master (us) | This PR (us) | Speedup |
|--------|-----------------|------------|-------------|--------------|---------|
| CUDA   | (1, 129, 33)    | 256        | 147         | 136          | 1.08    |
|        |                 | 128        | 153         | 128          | 1.20    |
|        | (100, 129, 20)  | 256        | 181         | 147          | 1.23    |
|        |                 | 128        | 171         | 137          | 1.25    |
|        | (1000, 129, 10) | 256        | 681         | 443          | 1.55    |
|        |                 | 128        | 632         | 446          | 1.42    |
| CPU    | (1, 129, 33)    | 256        | 106         | 104          | 1.02    |
|        |                 | 128        | 103         | 81           | 1.27    |
|        | (100, 129, 20)  | 256        | 2400        | 399          | 6.02    |
|        |                 | 128        | 2150        | 313          | 6.87    |
|        | (1000, 129, 10) | 256        | 13800       | 3740         | 3.69    |
|        |                 | 128        | 12700       | 2110         | 6.02    |

peterbell10 committed:
istft: Use unfold_backward instead of col2im

`unfold_backward` implements the same operation as `col2im` but without support
for 2d kernels or dilation. However, `istft` doesn't use any of those features
and `unfold_backward` actually has a faster `TensorIterator` based
implementation so we should use it here instead.

In the example from #87353 I see a 2x speedup on both CPU and CUDA.

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -1112,26 +1112,23 @@ Tensor istft(const Tensor& self, const int64_t n_fft, const optional<int64_t> ho
+    /*dim=*/1,
+    /*size=*/n_fft,
+    /*step=*/hop_length);
+  window_tmp = window_tmp.pow(2).expand({1, n_frames, n_fft});  // size: (1, n_fft, n_frames)
```
The comments is not accurate anymore?

albanD reviewed (approved):
Sounds good!
Just a small question about comments.

peterbell10 committed:
Fix size comments on "istft: Use unfold_backward instead of col2im"


`unfold_backward` implements the same operation as `col2im` but without support
for 2d kernels or dilation. However, `istft` doesn't use any of those features
and `unfold_backward` actually has a faster `TensorIterator` based
implementation so we should use it here instead.

In the example from #87353 I see a 2x speedup on both CPU and CUDA.

On a wider variety of sizes and inputs I still see speedups across the board, especially 
on CPU since `col2im` isn't parallelized but `unfold_backward` is:

| device | shape           | hop_length | Master (us) | This PR (us) | Speedup |
|--------|-----------------|------------|-------------|--------------|---------|
| CUDA   | (1, 129, 33)    | 256        | 147         | 136          | 1.08    |
|        |                 | 128        | 153         | 128          | 1.20    |
|        | (100, 129, 20)  | 256        | 181         | 147          | 1.23    |
|        |                 | 128        | 171         | 137          | 1.25    |
|        | (1000, 129, 10) | 256        | 681         | 443          | 1.55    |
|        |                 | 128        | 632         | 446          | 1.42    |
| CPU    | (1, 129, 33)    | 256        | 106         | 104          | 1.02    |
|        |                 | 128        | 103         | 81           | 1.27    |
|        | (100, 129, 20)  | 256        | 2400        | 399          | 6.02    |
|        |                 | 128        | 2150        | 313          | 6.87    |
|        | (1000, 129, 10) | 256        | 13800       | 3740         | 3.69    |
|        |                 | 128        | 12700       | 2110         | 6.02    |

[ghstack-poisoned]
